### ISSUE TITLE
Fuzy find space separated words

### ISSUE THREAD

**Issue by Ram-Z** [junegunn/fzf#178]
i.e. a mode where
```bash
echo "foo bar\nfoobar" | fzf -f "oobar"
```
would  not match `foo bar`, because the fuzy find would cross a space.

**Comment by junegunn** [comment]
I'm not convinced we need it. Can you give me a realistic example for which you want the option?
- Both `foobar` and `foo bar` match, but `foobar` will be given higher rank and appear before the latter. I mean it's already easier to pick that one, so why we would need another option? Maybe because you want to use `fzf` as a non-interactive filter?
- You might also want to consider `--extended-exact` (or `-e`)
- If the number of fields on each line is known beforehand, we can consider using `--nth` option
    - e.g. `fzf -n1,2,3 -f "oobar"`

**Comment by Ram-Z** [comment]
I thought of this when it took me more effort than expected when searching through my zsh history (`^R`). I usually know what command I'm looking for, so will search for that first. I also know more or less how the words are separated, so putting an explicit space between words helps filtering lots of stuff.

The history search uses the `--no-sort` option, therefore ignores the rank. I tried it with the rank, but it just felt wrong to have the history not sorted in chronological order.

My initial example was also incomplete. If there is a space in the search string it should be respected, i.e.

```bash
echo "foobar" | fzf -f "oo bar"
```
should not match, as you are explicitly searching for two words. One containing `oo` and the second `bar`.

Even better would be to be able to select the 'hard' separators, then for instance if you're searching through a long list of long paths and set the separator to `/`, you could search for `/a/b/abc` and be sure that you would only see paths that have an `a` in the first folder, a `b` in the second, and abc in the rest of the path.

**Comment by junegunn** [comment]
> The history search uses the --no-sort option, therefore ignores the rank. I tried it with the rank, but it just felt wrong to have the history not sorted in chronological order.

By the way, have you tried the updated version of `CTRL-R`, you can now toggle sort with `CTRL-R` key as mentioned in [the README page](https://github.com/junegunn/fzf#key-bindings-for-command-line).

**Comment by junegunn** [comment]
> Even better would be to be able to select the 'hard' separators, then for instance if you're searching through a long list of long paths and set the separator to /, you could search for /a/b/abc and be sure that you would only see paths that have an a in the first folder, a b in the second, and abc in the rest of the path.

I see what you mean, but it sounds too specific, and I'm going to have a hard time explaining it to an ordinary user of fzf :) And technically, it's how fzf currently works when `--sort` is turned on, as fzf prefers the items with shorter matched substrings.

**Comment by wellle** [comment]
I'll have to admit that I like the idea of hard separators. Especially in `^R` I've been in that situation many times.

The chronological ordering is very useful there, but it allows a lot of noise (very long matches) that I need to skip over manually most of the times.

**Comment by junegunn** [comment]
@wellle @Ram-Z 
So if I understood correctly, `aa/bb/cc` effectively translates to `a[^/]*?a[^/]*?/b[^/]*?b[^/]*?/c[^/]*?c[^/]*?`, and `aaa/bbb/ccc` matches the pattern but `aaa/bbb/xxx/ccc` does not. Is it correct? Or am I missing something? If that's the case, it might be useful if the input only consists of paths, but I don't think it's ideal for command history. For example, consider input string `cp hello/world aaa/bbb/ccc`, and you type in `cpaa/bb/cc` but it doesn't match because of `hello/world` before `aaa/`, and you're obliged to include `h/w` to the pattern.

**Comment by wellle** [comment]
Yeah I see your point. Maybe `'oobar` in `--extended` search mode is the better solution here. Because that way you can mark explicitly which parts of your query should fuzzy over spaces and what shouldn't.

**Comment by Ram-Z** [comment]
> So if I understood correctly, `aa/bb/cc` effectively translates to `a[^/]*?a[^/]*?/b[^/]*?b[^/]*?/c[^/]*?c[^/]*?,` and `aaa/bbb/ccc` matches the pattern but `aaa/bbb/xxx/ccc` does not. Is it correct?

Not quite, when a `/` appears in the search, it means that the part before and the part after must have at least one `/` between them. And the separator is not allowed if it wasn't explicitly added to the search pattern.

Rather something like `a[^/]*?a.*?/.*?b[^/]*?b.*?/.*?c[^/]*?c`. So for the string `aaa/bbb/xxx/ccc`

```
matched chars in uppercase
aa/bb/cc        MATCH     AAa/BBb/xxx/CCc
aa/bb/x/cc      MATCH     AAa/BBb/Xxx/CCc
aa/bb/xcc       NO MATCH  there is a / between x and c which is not explicitly searched
aa/bbx/cc       NO MATCH  there is a / between b and x which is not explicitly searched
```

> If that's the case, it might be useful if the input only consists of paths, but I don't think it's ideal for command history.

Indeed, if the hard separator is a `/` it makes only sense for paths, i.e. <kbd>Alt+C</kbd>.

For command history, hard separators on `space` makes a lot more sense than on `/`. `cp aa/bb/cc` would match `cp hello/world aaa/bbb/ccc` but not `cat p aaa/bbb/ccc` because `c` and `p` are separated by a space.

> Yeah I see your point. Maybe `'oobar` in `--extended` search mode is the better solution here. Because that way you can mark explicitly which parts of your query should fuzzy over spaces and what shouldn't.

The problem here is that it doesn't fuzzy anymore, i.e. `'obr` would not match `foobar`

**Comment by junegunn** [comment]
I see. It makes sense. But I must say it's a bit hard to follow at first. The problem with the suggested approach is that when space is used as a hard separator, it is no longer compatible with the current extended-search mode, where spaces are used to separate multiple search terms. Which means in order to implement the scheme, we cannot augment the preexisting modes with the concept of hard separator but instead have to add another mode, so we end up with 4 different search modes, and I'd really like to avoid that.

I suggest that you try `CTRL-R` for enabling sort (it was added recently) and "quoting" in extended-search mode for a while. I'm sure they will help you a lot in most of the cases. I'll leave this issue open. Feel free to share more of your ideas. Thanks.

**Comment by Ram-Z** [comment]
Indeed, it's not compatible with extended-search mode. I think it could be an extension of the normal search mode though, it could even be integrate it into the main search loop [algo.go#L34](https://github.com/junegunn/fzf/blob/master/src/algo/algo.go#L34) I believe.

Another approach would be to consider each string separated by a hard separator as a different pattern to search, with the restriction that the matched string does not contain the hard separator. I this case the order is not respected anymore. `aa/bb` would match `AAa/BaB` aswell as `BaB/AAa` but not `ba/aba`. It's still not compatible with extended-search when the hard separator is a space though.

> I suggest that you try <kbd>CTRL-R</kbd> for enabling sort (it was added recently) and "quoting" in extended-search mode for a while. I'm sure they will help you a lot in most of the cases.

I have for the past few days, it works somewhat better than with `--no-sort`, but tends to fetch similar commands from far back in the past which might have a better rank but have no relations to what I'm currently doing. So I still need to type more than I think should be necessary. I'll try it for a bit longer though, maybe I'll get accustomed to it.
